### PR TITLE
AVR-DES + DES Refactoring #3499

### PR DESCRIPTION
**Pull request by wargio**
I've done some refactoring on the DES crypto in order to fix avr des.

Now since there are no example of the problem (@condret help me please) i suggest to wait a regression test for r2 test suite.

### CODE REVIEW

**Comment by condret** [comment]
first thing that i recognize is that someone commented this https://github.com/wargio/radare2/blob/ad4e566e0e9036e201a829dc07a2b3f1111ebd70/libr/anal/p/anal_avr.c#L615 out. would be great if you can fix it, so that the handler actually gets called.

Did you check your DES-implementation against testvectors, such as these https://www.cosic.esat.kuleuven.be/nessie/testvectors/bc/des/Des-64-64.test-vectors ?

Your sboxes, well you are not using the same sboxes as flanfly and I used, when we created this DES. We copied them from the book "Understanding Cryptography". I know for sure that these sboxes worked in a DES-implementation I created once, and then accidentially deleted.

If you've tested your implementation and confirm that it works correctly, I think i'll be fine

**Comment by wargio** [comment]
The problem isn't on my des implementation, which is correct (yes, test vectors were my initial test,before importing my code to r2), but the avr one. Ok for the handler, but is there a bin with the des code that I can actually test? One that reproduce the bug

File: libr/anal/p/anal_avr.c [hunk 1]
```diff
@@ -1564,21 +1566,46 @@ static int avr_op(RAnal *anal, RAnalOp *op, ut64 addr, const ut8 *buf, int len)
...
-	r = (int)key;
-	r_anal_esil_reg_read (esil, "hf", &key, NULL);
-	enc = (int)key;
+	r_anal_esil_reg_read (esil, "hf", &encrypt, NULL);
 	r_anal_esil_reg_read (esil, "deskey", &key, NULL);
 	r_anal_esil_reg_read (esil, "text", &text, NULL);
-	key = r_des_get_roundkey (key, r, enc);
-	text = r_des_round (text, key);
+
+	key_lo = key & 0xffffffff;
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: use UT32_MAX

File: libr/anal/p/anal_avr.c [hunk 2]
```diff
@@ -1564,21 +1566,46 @@ static int avr_op(RAnal *anal, RAnalOp *op, ut64 addr, const ut8 *buf, int len)
...
+	key_hi = key >> 32;
+	buf_lo = text & 0xffffffff;
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: same here

File: libr/util/des.c [hunk 3]
```diff
@@ -1,249 +1,249 @@
...
-	const ut32 *p = (inv ? INV_IP : IP);
-
-	while(bit < 64) {
-		ret |= (state & (1ul << (p[bit] - 1)) ? 1ul << bit : 0);
-		bit += 1;
+// keylo & keyhi are the derivated round keys
+// deskeylo & deskeyhi are the des derivated keys
+R_API void r_des_round_key (int i, ut32 *keylo, ut32 *keyhi, ut32 *deskeylo, ut32 *deskeyhi) {
+	ut32 deslo, deshi;
+	if (!keylo || !keyhi || !deskeylo || !deskeyhi) return;
```
Comment (radare) [inline libr/util/des.c]: add braces

File: libr/util/des.c [hunk 4]
```diff
@@ -1,249 +1,249 @@
...
-
-	while(bit < 48) {
-		ret |= (half & (1ul << (E[bit] - 1)) ? 1ul << bit : 0);
-		bit += 1;
-	}
-
-	return ret;
+R_API void r_des_round (ut32 *buflo, ut32 *bufhi, ut32 *roundkeylo, ut32 *roundkeyhi) {
+	ut32 perm, lo, hi;
+	if (!buflo || !bufhi || !roundkeylo || !roundkeyhi) return;
```
Comment (radare) [inline libr/util/des.c]: add braces

File: libr/util/des.c [hunk 5]
```diff
@@ -1,249 +1,249 @@
...
+	lo = *buflo;
```
Comment (radare) [inline libr/util/des.c]: define ut32 lo here, no need to separate definition from assignment

**Comment by radare** [comment]
can you provide a test?

**Comment by wargio** [comment]
i would love to, but i'm waiting @condret for that. he created the issue that i'm trying to handle

**Comment by radare** [comment]
doesnt builds

€˜const char *â€™
/home/travis/build/radare/radare2/libr/..//libr/anal/p/anal_avr.c: In function â€˜_inst__desâ€™:
/home/travis/build/radare/radare2/libr/..//libr/anal/p/anal_avr.c:619:39: error: â€˜dâ€™ undeclared (first use in this function)
/home/travis/build/radare/radare2/libr/..//libr/anal/p/anal_avr.c:619:39: note: each undeclared identifier is reported only once for each function it appears in
make[3]: *** [/home/travis/

File: libr/anal/p/anal_avr.c [hunk 6]
```diff
@@ -12,6 +12,8 @@
 #include <r_asm.h>
 #include <r_anal.h>
 
+RDESContext desctx;
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: static maybe?

File: libr/anal/p/anal_avr.c [hunk 7]
```diff
@@ -1564,21 +1566,50 @@ static int avr_op(RAnal *anal, RAnalOp *op, ut64 addr, const ut8 *buf, int len)
...
 static int avr_custom_des (RAnalEsil *esil) {
-	ut64 key, text;
-	int r, enc;
+	ut64 key, encrypt, text,des_round;
+	ut32 key_lo, key_hi, buf_lo, buf_hi;
 	if (!esil || !esil->anal || !esil->anal->reg) {
 		return false;
 	}
-	if (!__esil_pop_argument(esil, &key)) {
+	if (!__esil_pop_argument(esil, &des_round)) {
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: use r2-indent or add a space before (

File: libr/anal/p/anal_avr.c [hunk 8]
```diff
@@ -1564,21 +1566,50 @@ static int avr_op(RAnal *anal, RAnalOp *op, ut64 addr, const ut8 *buf, int len)
...
+
+	if (des_round != desctx.round) {
+		desctx.round = des_round;
+	}
+	
+	if (!desctx.round) {
+		int i;
+		//generating all round keys
+		r_des_permute_key (&key_lo, &key_hi);
+		for (i = 0; i < 16; ++i) {
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: dont use preincrement if its not strictly necessary

File: libr/anal/p/anal_avr.c [hunk 9]
```diff
@@ -1724,6 +1755,7 @@ static int esil_avr_hook_reg_write(RAnalEsil *esil, const char *name, ut64 *val)
 static int esil_avr_init(RAnalEsil *esil) {
 	if (!esil)
```
Comment (radare) [inline libr/anal/p/anal_avr.c]: missing braces

File: libr/util/des.c [hunk 10]
```diff
@@ -1,249 +1,227 @@
...
+	hi ^= perm; lo ^= perm << 4;
+	perm = ((lo >> 16) ^ hi) & 0x0000FFFF;
+	hi ^= perm; lo ^= perm << 16;
+	perm = ((hi >> 2) ^ lo) & 0x33333333;
+	lo ^= perm; hi ^= perm << 2;
+	perm = ((hi >> 8) ^ lo) & 0x00FF00FF;
+	lo ^= perm; hi ^= perm << 8;
+	perm = ((lo >> 1) ^ hi) & 0x55555555;
+	hi ^= perm; lo ^= perm << 1;
+	*blocklo = ROTL(lo, 1);
```
Comment (radare) [inline libr/util/des.c]: use r2-indent or add spaces before (

File: libr/util/des.c [hunk 11]
```diff
@@ -1,249 +1,227 @@
...
-
-R_API ut32 r_des_p (ut32 half) {
-	ut32 ret = 0;
-	ut32 bit = 0;
-
-	while(bit < 32) {
-		ret |= (half & (1ul << (P[bit] - 1)) ? 1ul << bit : 0);
-		bit += 1;
+	if (i == 0 || i == 1 || i == 8 || i == 15) {
+		*deskeylo = ROTL28(*deskeylo, 1);
```
Comment (radare) [inline libr/util/des.c]: missing spaces

File: libr/util/des.c (hunk 10 above)
Comment (wargio) [inline libr/util/des.c]: it's a macro

**Pull request merged**